src/dso_api/dynamic_api/permissions.py

```diff
@@ -58,19 +55,28 @@ def fetch_scopes_for_model(model: Type[Model]) -> TableScopes:
     return fetch_scopes_for_dataset_table(dataset, table)
 
 
-def get_unauthorized_fields(request, model) -> set:
-    """Check which field names should be excluded"""
+def get_unauthorized_fields(request, model) -> Set[str]:
+    """Returns a list of field names to be excluded from a response."""
+    # is_authorized_for is added by authorization_django middleware
+    # only if an authorization check is necessary.
+    if not hasattr(request, "is_authorized_for"):
+        return set()
+
     model_scopes = fetch_scopes_for_model(model)
     unauthorized_fields = set()
-    # is_authorized_for is added by authorization_django middleware
-    if hasattr(request, "is_authorized_for"):
-        for model_field in model._meta.get_fields():
-            scopes = model_scopes.fields.get(model_field.name)
-            scopes = scopes.union(model_scopes.table) if scopes else model_scopes.table
+
+    for model_field in model._meta.get_fields():
+        name = model_field.name
+
+        required = set()
+        if (field_scope := model_scopes.fields.get(name)) is not None:
+            required.add(field_scope)
+        required |= model_scopes.table
```

Comment:
This does exactly same as following did before, is there real reason for rewrite?

```
scopes = model_scopes.fields.get(model_field.name)
scopes = scopes.union(model_scopes.table) if scopes else model_scopes.table
```

Reply:
Or `required = model_scopes.fields.get(model_field.name, set()) | model_scopes.table`

Reply:
The change is necessary because `model_scopes.fields[name]` is now a `str` instead of the previous `set`. I couldn't come up with an elegant one-liner to create a singleton set for a `str` and an empty set for `None`.

Reply:
Aha, I see your point.

Reply:
You could do this: `required = set(model_scopes.fields.get(model_field.name)) - {None} | model_scopes.table`, but I doubt if that's more readable :-)

Reply:
I could write

```py
required = set(s for s in [model_scopes.fields.get(model_field.name)] if s is not None) | model_scopes.table
```

but I'm not sure if that's any better. Maybe

```py
required = set([field_scope]) if (field_scope := model_scopes.fields.get(name)) is not None else set()
required |= model_scopes.table
```

? That's hardly less complicated than the current code.

Reply:
Following is much more readable, also makes it clear that field scope is singular.
```
 required = model_scopes.table
 field_scope = model_scopes.fields.get(model_field.name)
 if field_scope:
     required.union(set(field_scope))
```


It's nice that `TableScopes` was changed to present fields as dict with strings as values, but then why do we still have table scopes as set, isn't it more confusing now?

Reply:
I'll separate the assignment out (the new Python syntax is strange anyway).

The rest of your suggestion is pretty much what I started with. It doesn't work because (1) `set(field_scope)` constructs a set of the characters of `field_scope` (very hard bug to catch) and (2) `union` doesn't modify the left-hand object. `set.add` avoids both problems and is explicit.

`TableScopes.table` is a set because it stores the table scope and/or the dataset scope. I could separate out the dataset scope, but this was already the case.

Reply:
Oh, sorry for brainfart code :) 

replace: `required.union(set(field_scope))` with `required = required.union([field_scope])`

Reply:
Oh I didn't want to start with `required = model_scopes.table` because that makes it easy to accidentally modify the `TableScopes.table` set in-place. Maybe we should make that a `frozenset`? It's not a big deal now, but it's better to be safe than sorry...

Reply:
I rewrote it following your suggestion **and** made the set a frozenset.